### PR TITLE
48523_course_task_manipulation

### PR DESCRIPTION
**Pull request by noor-framgia**

### CODE REVIEW

File: app/views/supervisor/course_subjects/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        <%= hidden_field_tag :course_id, @course.id %>
+        <% for subject in @course.subjects do %>
+          <%= subject.name %><br/>
+          <% for task in subject.tasks do %>
+            <%= task.name %>
+            <%= check_box_tag "task_ids[]", task.id  %><br/>
+          <% end %>
+        <% end %>
+
+        <%= submit_tag 'Modify tasks', class: "btn btn-primary" %>
```
Comment (shakil-fg) [inline app/views/supervisor/course_subjects/edit.html.erb]: Double quote

File: app/views/supervisor/course_subjects/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+<div class="row">
+  <div class="panel panel-default panel-box">
+    <div class="panel-heading panel-size-margin">
+      <%= @course.name %>
+    </div>
+  </div>
+
+  <div class="row">
+    <div class="col-md-8 col-md-offset-2">
+      <%= form_for @course_subjects, :url => {controller: 'supervisor/course_subjects', action: 'update'}, method: :patch do %>
```
Comment (shakil-fg) [inline app/views/supervisor/course_subjects/edit.html.erb]: Double quote

**Comment by manazirahsan** [comment]
okey

File: app/models/course_subject.rb [hunk 3]
```diff
@@ -3,4 +3,10 @@ class CourseSubject < ActiveRecord::Base
   belongs_to :subject
 
   validates :course_id, uniqueness: {scope: :subject_id}
+
+  class << self
+    def find_by_course_id_with_subject_id course_id, subject_id
+      CourseSubject.find_by(course_id: course_id, subject_id: subject_id)
```
Comment (MuhammadTamzid) [inline app/models/course_subject.rb]: remove ()
CourseSubject.find_by course_id: course_id, subject_id: subject_id

File: app/views/supervisor/courses/show.html.erb [hunk 4]
```diff
@@ -11,7 +11,10 @@
         <%= link_to t(:edit), edit_supervisor_course_path(@course), class: "btn btn-warning" %>
         <% unless @course.close? || @course.continue? %>
           <%= link_to t(:close), supervisor_course_path(@course, type: "close"), method: :patch,
-                      data: {confirm: t(:confirm_message_close)}, class: "btn btn-warning" %>
+                      data: {confirm: t(:confirm_message_close)}, class: "btn btn-danger" %>
+          <% end %>
+          <% unless @course.close? %>
+          <%= link_to t(:manipulate_task), edit_supervisor_course_subject_path(@course), class: "btn btn-info" %>
         <% end %>
```
Comment (MuhammadTamzid) [inline app/views/supervisor/courses/show.html.erb]: <% unless @course.close? || @course.continue? %>
  <%= link_to t(:close), supervisor_course_path(@course, type: "close"), method: :patch,
    data: {confirm: t(:confirm_message_close)}, class: "btn btn-danger" %>
<% end %>
<% unless @course.close? %>
  <%= link_to t(:manipulate_task), edit_supervisor_course_subject_path(@course), class: "btn btn-info" %>
<% end %>

File: app/views/supervisor/courses/show.html.erb [hunk 5]
```diff
@@ -11,7 +11,10 @@
         <%= link_to t(:edit), edit_supervisor_course_path(@course), class: "btn btn-warning" %>
         <% unless @course.close? || @course.continue? %>
           <%= link_to t(:close), supervisor_course_path(@course, type: "close"), method: :patch,
-                      data: {confirm: t(:confirm_message_close)}, class: "btn btn-warning" %>
+                      data: {confirm: t(:confirm_message_close)}, class: "btn btn-danger" %>
+          <% end %>
```
Comment (MuhammadTamzid) [inline app/views/supervisor/courses/show.html.erb]: indentation

File: app/views/supervisor/courses/show.html.erb [hunk 6]
```diff
@@ -11,7 +11,10 @@
...
+          <% unless @course.close? %>
```
Comment (MuhammadTamzid) [inline app/views/supervisor/courses/show.html.erb]: indentation

File: app/views/supervisor/courses/show.html.erb (hunk 5 above)
Comment (noor-framgia) [inline app/views/supervisor/courses/show.html.erb]: breaks 80 char. so wrote in 2nd line

File: app/controllers/supervisor/course_subjects_controller.rb [hunk 7]
```diff
@@ -1,2 +1,36 @@
...
+  end
+
+  def edit
+    @course= Course.find params[:id]
+    @course_subjects = CourseSubject.new
+  end
+
+  def update
+    params[:task_ids].each do |task_id|
+      CourseSubjectTask.create(course_subject:
```
Comment (dieunb) [inline app/controllers/supervisor/course_subjects_controller.rb]: I don't understand why you can implement as it. For update action, you can find course_subject by 
```
@course_subject = CourseSubject.find params[:id]
```

Should update course_subject, and nested to update course_subject_task

File: app/controllers/supervisor/course_tasks_controller.rb [hunk 8]
```diff
@@ -0,0 +1,3 @@
+class Supervisor::CourseTasksController < ApplicationController
```
Comment (dieunb) [inline app/controllers/supervisor/course_tasks_controller.rb]: i think, we no need use this controller

Comment (noor-framgia) [inline app/controllers/supervisor/course_tasks_controller.rb]: yes, i forgot to remove, sorry

File: app/views/supervisor/course_subjects/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,14 @@
+<div class="row">
+  <div class="col-md-8 col-md-offset-2">
+
+      <% for subject in @course.subjects do %>
```
Comment (dieunb) [inline app/views/supervisor/course_subjects/show.html.erb]: should use `each` to loop, http://graysoftinc.com/early-steps/the-evils-of-the-for-loop

File: app/views/supervisor/courses/show.html.erb [hunk 10]
```diff
@@ -11,7 +11,10 @@
         <%= link_to t(:edit), edit_supervisor_course_path(@course), class: "btn btn-warning" %>
         <% unless @course.close? || @course.continue? %>
           <%= link_to t(:close), supervisor_course_path(@course, type: "close"), method: :patch,
-                      data: {confirm: t(:confirm_message_close)}, class: "btn btn-warning" %>
+            data: {confirm: t(:confirm_message_close)}, class: "btn btn-danger" %>
+          <% end %>
+          <% unless @course.close? %>
+          <%= link_to t(:manipulate_task), edit_supervisor_course_subject_path(@course), class: "btn btn-info" %>
```
Comment (dieunb) [inline app/views/supervisor/courses/show.html.erb]: indent

File: db/schema.rb [hunk 11]
```diff
@@ -11,7 +11,11 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
+<<<<<<< HEAD
```
Comment (dieunb) [inline db/schema.rb]: conflicted code

File: app/controllers/supervisor/course_subjects_controller.rb [hunk 12]
```diff
@@ -1,2 +1,46 @@
 class Supervisor::CourseSubjectsController < ApplicationController
+  before_action :load_subject
+  before_action :load_task
+  before_action :load_course
+
+  def show
+    @course= Course.find params[:id]
```
Comment (dieunb) [inline app/controllers/supervisor/course_subjects_controller.rb]: i think 
```
@course_subject = CourseSubject.find params[:id]
@course = @course_subject.course
```

File: app/controllers/supervisor/course_subjects_controller.rb [hunk 13]
```diff
@@ -1,2 +1,46 @@
...
+  end
+
+  def new
+
+    @course= Course.find params[:id]
```
Comment (dieunb) [inline app/controllers/supervisor/course_subjects_controller.rb]: i don't understand `params[:id]`, where is it? It mean, params[:course_id]

File: app/controllers/supervisor/course_subjects_controller.rb [hunk 14]
```diff
@@ -1,2 +1,46 @@
...
+    else
+      redirect_to supervisor_course_subject_url
+    end
+  end
+
+  private
+
+  def course_subject_params
+    params.require(:course_subject).permit :id, :course_id,
+      course_subject_tasks_attributes: [:course_subject_id,'task_ids[]']
```
Comment (dieunb) [inline app/controllers/supervisor/course_subjects_controller.rb]: `'task_ids[]'`, don't understand!

File: app/models/course_subject.rb [hunk 15]
```diff
@@ -1,6 +1,15 @@
...
   belongs_to :course
   belongs_to :subject
+  has_many :course_subject_tasks
 
   validates :course_id, uniqueness: {scope: :subject_id}
+
+  accepts_nested_attributes_for :course_subject_tasks  , allow_destroy: true
+
+  class << self
+    def find_by_course_id_with_subject_id course_id, subject_id
```
Comment (dieunb) [inline app/models/course_subject.rb]: is it necessary?

File: app/views/supervisor/course_subjects/new.html.erb [hunk 16]
```diff
@@ -0,0 +1,30 @@
+<% provide :title, t(:update_course)%>
```
Comment (dieunb) [inline app/views/supervisor/course_subjects/new.html.erb]: course_subject was created when assign subject to course, i think it isn't necessary.

File: app/models/course_subject.rb (hunk 15 above)
Comment (noor-framgia) [inline app/models/course_subject.rb]: as i implemented previously, i was in work.

File: app/views/supervisor/course_subjects/new.html.erb (hunk 16 above)
Comment (noor-framgia) [inline app/views/supervisor/course_subjects/new.html.erb]: its actually task modification, supervisor can modify (add more task to a trainee after assigning)

File: app/controllers/supervisor/course_subjects_controller.rb (hunk 13 above)
Comment (noor-framgia) [inline app/controllers/supervisor/course_subjects_controller.rb]: <% unless @course.close? %>
          <%= link_to t(:manipulate_task),  new_supervisor_course_subject_path(:id => @course.id),class: "btn btn-info" %>
        <% end %>

Comment (noor-framgia) [inline app/controllers/supervisor/course_subjects_controller.rb]: in course/show.html.erb

File: app/controllers/supervisor/course_subjects_controller.rb [hunk 17]
```diff
@@ -1,2 +1,47 @@
...
+
+  def new
+    @course= Course.find params[:id]
+    @course_subject = CourseSubject.new
+    @course_subject.course_subject_tasks.build
+
+  end
+
+  def create
+    debugger
```
Comment (dieunb) [inline app/controllers/supervisor/course_subjects_controller.rb]: pls, remove breack point

File: app/views/supervisor/course_subjects/show.html.erb [hunk 18]
```diff
@@ -0,0 +1,14 @@
+<div class="row">
+  <div class="col-md-8 col-md-offset-2">
+
+      <% @course.subjects.each do |subject| %>
+        <strong><%= subject.name %><br/></strong>
+          <div>
+            <% for task in subject.tasks do %>
```
Comment (dieunb) [inline app/views/supervisor/course_subjects/show.html.erb]: use each instead for

File: app/views/supervisor/courses/show.html.erb [hunk 19]
```diff
@@ -11,7 +11,10 @@
         <%= link_to t(:edit), edit_supervisor_course_path(@course), class: "btn btn-warning" %>
         <% unless @course.close? || @course.continue? %>
           <%= link_to t(:close), supervisor_course_path(@course, type: "close"), method: :patch,
-                      data: {confirm: t(:confirm_message_close)}, class: "btn btn-warning" %>
+            data: {confirm: t(:confirm_message_close)}, class: "btn btn-danger" %>
+          <% end %>
+        <% unless @course.close? %>
+          <%= link_to t(:manipulate_task),  new_supervisor_course_subject_path(:id => @course.id),class: "btn btn-info" %>
```
Comment (dieunb) [inline app/views/supervisor/courses/show.html.erb]: Convention

File: app/controllers/supervisor/course_subjects_controller.rb (hunk 17 above)
Comment (noor-framgia) [inline app/controllers/supervisor/course_subjects_controller.rb]: oh  so sorry

File: app/views/supervisor/courses/show.html.erb (hunk 19 above)
Comment (noor-framgia) [inline app/views/supervisor/courses/show.html.erb]: is it ok now?

**Pull request merged**